src/cargo/ops/cargo_output_metadata.rs

```diff
@@ -39,10 +39,47 @@ pub fn output_metadata<'a>(opt: OutputMetadataOptions,
 #[derive(RustcEncodable)]
 pub struct ExportInfo {
     packages: Vec<Package>,
-    resolve: Resolve,
+    resolve: MetadataResolve,
     version: u32,
 }
 
+struct MetadataResolve(Resolve);
+
+impl Encodable for MetadataResolve {
+    fn encode<S: Encoder>(&self, s: &mut S) -> Result<(), S::Error> {
+        EncodableResolve::new(&self.0).encode(s)
+    }
+}
+
+#[derive(RustcEncodable)]
+struct EncodableResolve<'a> {
+    root: &'a PackageId,
+    nodes: Vec<Node<'a>>,
+}
+
+#[derive(RustcEncodable)]
+struct Node<'a> {
+    id: &'a PackageId,
+    dependencies: Vec<&'a PackageId>,
+}
+
+impl<'a> EncodableResolve<'a> {
+    fn new(resolve: &Resolve) -> EncodableResolve {
+        EncodableResolve {
+            root: resolve.root(),
+            nodes: resolve.iter()
+                          .map(|id| {
+                              Node {
+                                  id: id,
+                                  dependencies: resolve.deps(id)
+                                                       .map(|it| it.collect())
+                                                       .unwrap_or(Vec::new()),
+                              }
+                          })
+                          .collect(),
```

Comment:
Stylistically this tends to look moreso like:

```rust
nodes: resolve.iter().map(|id| {
    Node {
        foo: ...
    }
}).collect()
```

(helps cut down on the indentation